<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta http-equiv="X-UA-Compatible" content="ie=edge">
    <title>Recognition Site Modification</title>

    <!-- Font Icon -->
    <link rel="stylesheet" href="fonts/material-icon/css/material-design-iconic-font.min.css">
    <link rel="stylesheet" href="vendor/nouislider/nouislider.min.css">

    <!-- Main css -->
    <link rel="stylesheet" href="css/style.css">

    <!-- Pyscript -->
    <link rel="stylesheet" href="https://pyscript.net/alpha/pyscript.css"/>
    <script defer src="https://pyscript.net/alpha/pyscript.js"></script>
</head>
<body>

    <div class="main">

        <div class="container">
            <div class="signup-content">
                <div class="signup-img">
                    <img src="images/form-img3.jpeg" alt="">
                    <!-- <div class="signup-img-content">
                        <h2>Recognition Site Modifier </h2>
                        <p>while seats are available !</p>
                    </div> -->
                </div>
                <div class="signup-form">
                    <form onsubmit="return false" class="register-form" id="register-form">
                        <div>
                            <h2>Recognition Site Modifier </h2>
                        </div>
                        <div class="form-row">
                            <div class="form-group">
                                <div class="form-select">
                                    <label for="enzyme" class="required">Select an enzyme</label>
                                    <select id="enzyme">
                                        <option value="" disabled selected>Select an enzyme</option>
                                        <option value="0">AatII</option>
                                        <option value="1">AbaSI</option>
                                        <option value="2">Acc65I</option>
                                        <option value="3">AccI</option>
                                        <option value="4">AciI</option>
                                        <option value="5">AclI</option>
                                        <option value="6">AcuI</option>
                                        <option value="7">AfeI</option>
                                        <option value="8">AflII</option>
                                        <option value="9">AflIII</option>
                                        <option value="10">AgeI-HF�</option>
                                        <option value="11">AhdI</option>
                                        <option value="12">AleI-v2</option>
                                        <option value="13">AluI</option>
                                        <option value="14">AlwI</option>
                                        <option value="15">AlwNI</option>
                                        <option value="16">ApaI</option>
                                        <option value="17">ApaLI</option>
                                        <option value="18">ApeKI</option>
                                        <option value="19">ApoI-HF�</option>
                                        <option value="20">AscI</option>
                                        <option value="21">AseI</option>
                                        <option value="22">AsiSI</option>
                                        <option value="23">AvaI</option>
                                        <option value="24">AvaII</option>
                                        <option value="25">AvrII</option>
                                        <option value="26">BaeGI</option>
                                        <option value="27">BaeI</option>
                                        <option value="28">BamHI</option>
                                        <option value="29">BamHI-HF�</option>
                                        <option value="30">BanI</option>
                                        <option value="31">BanII</option>
                                        <option value="32">BbsI</option>
                                        <option value="33">BbsI-HF�</option>
                                        <option value="34">BbvCI</option>
                                        <option value="35">BbvI</option>
                                        <option value="36">BccI</option>
                                        <option value="37">BceAI</option>
                                        <option value="38">BcgI</option>
                                        <option value="39">BciVI</option>
                                        <option value="40">BclI</option>
                                        <option value="41">BclI-HF</option>
                                        <option value="42">BcoDI</option>
                                        <option value="43">BfaI</option>
                                        <option value="44">BfuAI</option>
                                        <option value="45">BglI</option>
                                        <option value="46">BglII</option>
                                        <option value="47">BlpI</option>
                                        <option value="48">BmgBI</option>
                                        <option value="49">BmrI</option>
                                        <option value="50">BmtI-HF�</option>
                                        <option value="51">BpmI</option>
                                        <option value="52">Bpu10I</option>
                                        <option value="53">BpuEI</option>
                                        <option value="54">BsaAI</option>
                                        <option value="55">BsaBI</option>
                                        <option value="56">BsaHI</option>
                                        <option value="57">BsaI</option>
                                        <option value="58">BsaI-HF�v2</option>
                                        <option value="59">BsaJI</option>
                                        <option value="60">BsaWI</option>
                                        <option value="61">BsaXI</option>
                                        <option value="62">BseRI</option>
                                        <option value="63">BseYI</option>
                                        <option value="64">BsgI</option>
                                        <option value="65">BsiEI</option>
                                        <option value="66">BsiHKAI</option>
                                        <option value="67">BsiWI�</option>
                                        <option value="68">BsiWI-HF�</option>
                                        <option value="69">BslI</option>
                                        <option value="70">BsmAI</option>
                                        <option value="71">BsmBI-v2</option>
                                        <option value="72">BsmFI</option>
                                        <option value="73">BsmI</option>
                                        <option value="74">BsoBI</option>
                                        <option value="75">Bsp1286I</option>
                                        <option value="76">BspCNI</option>
                                        <option value="77">BspDI</option>
                                        <option value="78">BspEI</option>
                                        <option value="79">BspHI</option>
                                        <option value="80">BspMI</option>
                                        <option value="81">BspQI</option>
                                        <option value="82">BsrBI</option>
                                        <option value="83">BsrDI</option>
                                        <option value="84">BsrFI-v2</option>
                                        <option value="85">BsrGI-HF�</option>
                                        <option value="86">BsrI</option>
                                        <option value="87">BssHII</option>
                                        <option value="88">BssSI-v2</option>
                                        <option value="89">BstAPI</option>
                                        <option value="90">BstBI</option>
                                        <option value="91">BstEII-HF�</option>
                                        <option value="92">BstNI</option>
                                        <option value="93">BstUI</option>
                                        <option value="94">BstXI</option>
                                        <option value="95">BstYI</option>
                                        <option value="96">BstZ17I-HF�</option>
                                        <option value="97">Bsu36I</option>
                                        <option value="98">BtgI</option>
                                        <option value="99">BtgZI</option>
                                        <option value="100">BtsCI</option>
                                        <option value="101">BtsIMutI</option>
                                        <option value="102">BtsI-v2</option>
                                        <option value="103">Cac8I</option>
                                        <option value="104">ClaI</option>
                                        <option value="105">CspCI</option>
                                        <option value="106">CviAII</option>
                                        <option value="107">CviKI-1</option>
                                        <option value="108">CviQI</option>
                                        <option value="109">DdeI</option>
                                        <option value="110">DpnI</option>
                                        <option value="111">DpnII</option>
                                        <option value="112">DraI</option>
                                        <option value="113">DraIII-HF�</option>
                                        <option value="114">DrdI</option>
                                        <option value="115">EaeI</option>
                                        <option value="116">EagI-HF�</option>
                                        <option value="117">EarI</option>
                                        <option value="118">EciI</option>
                                        <option value="119">Eco53kI</option>
                                        <option value="120">EcoNI</option>
                                        <option value="121">EcoO109I</option>
                                        <option value="122">EcoP15I</option>
                                        <option value="123">EcoRI�</option>
                                        <option value="124">EcoRI-HF�</option>
                                        <option value="125">EcoRV�</option>
                                        <option value="126">EcoRV-HF�</option>
                                        <option value="127">Esp3I</option>
                                        <option value="128">FatI</option>
                                        <option value="129">FauI</option>
                                        <option value="130">Fnu4HI</option>
                                        <option value="131">FokI</option>
                                        <option value="132">FseI</option>
                                        <option value="133">FspI</option>
                                        <option value="134">HaeII</option>
                                        <option value="135">HaeIII</option>
                                        <option value="136">HgaI</option>
                                        <option value="137">HhaI</option>
                                        <option value="138">HincII</option>
                                        <option value="139">HindIII</option>
                                        <option value="140">HindIII-HF�</option>
                                        <option value="141">HinfI</option>
                                        <option value="142">HinP1I</option>
                                        <option value="143">HpaI</option>
                                        <option value="144">HpaII</option>
                                        <option value="145">HphI</option>
                                        <option value="146">Hpy166II</option>
                                        <option value="147">Hpy188I</option>
                                        <option value="148">Hpy188III</option>
                                        <option value="149">Hpy99I</option>
                                        <option value="150">HpyAV</option>
                                        <option value="151">HpyCH4III</option>
                                        <option value="152">HpyCH4IV</option>
                                        <option value="153">HpyCH4V</option>
                                        <option value="154">I-CeuI</option>
                                        <option value="155">I-SceI</option>
                                        <option value="156">KasI</option>
                                        <option value="157">KpnI-HF�</option>
                                        <option value="158">LpnPI</option>
                                        <option value="159">MboI</option>
                                        <option value="160">MboII</option>
                                        <option value="161">MfeI-HF�</option>
                                        <option value="162">MluCI</option>
                                        <option value="163">MluI-HF�</option>
                                        <option value="164">MlyI</option>
                                        <option value="165">MmeI</option>
                                        <option value="166">MnlI</option>
                                        <option value="167">MscI</option>
                                        <option value="168">MseI</option>
                                        <option value="169">MslI</option>
                                        <option value="170">MspA1I</option>
                                        <option value="171">MspI</option>
                                        <option value="172">MspJI</option>
                                        <option value="173">MwoI</option>
                                        <option value="174">NaeI</option>
                                        <option value="175">NarI</option>
                                        <option value="176">Nb.BbvCI</option>
                                        <option value="177">Nb.BsmI</option>
                                        <option value="178">Nb.BsrDI</option>
                                        <option value="179">Nb.BssSI</option>
                                        <option value="180">Nb.BtsI</option>
                                        <option value="181">NciI</option>
                                        <option value="182">NcoI�</option>
                                        <option value="183">NcoI-HF�</option>
                                        <option value="184">NdeI</option>
                                        <option value="185">NgoMIV</option>
                                        <option value="186">NheI-HF�</option>
                                        <option value="187">NlaIII</option>
                                        <option value="188">NlaIV</option>
                                        <option value="189">NmeAIII</option>
                                        <option value="190">NotI�</option>
                                        <option value="191">NotI-HF�</option>
                                        <option value="192">NruI-HF�</option>
                                        <option value="193">NsiI</option>
                                        <option value="194">NsiI-HF�</option>
                                        <option value="195">NspI</option>
                                        <option value="196">Nt.AlwI</option>
                                        <option value="197">Nt.BbvCI</option>
                                        <option value="198">Nt.BsmAI</option>
                                        <option value="199">Nt.BspQI</option>
                                        <option value="200">Nt.BstNBI</option>
                                        <option value="201">Nt.CviPII</option>
                                        <option value="202">PacI</option>
                                        <option value="203">PaeR7I</option>
                                        <option value="204">PaqCI�</option>
                                        <option value="205">PciI</option>
                                        <option value="206">PflFI</option>
                                        <option value="207">PflMI</option>
                                        <option value="208">PI-PspI</option>
                                        <option value="209">PI-SceI</option>
                                        <option value="210">PleI</option>
                                        <option value="211">PluTI</option>
                                        <option value="212">PmeI</option>
                                        <option value="213">PmlI</option>
                                        <option value="214">PpuMI</option>
                                        <option value="215">PshAI</option>
                                        <option value="216">PsiI-v2</option>
                                        <option value="217">PspGI</option>
                                        <option value="218">PspOMI</option>
                                        <option value="219">PspXI</option>
                                        <option value="220">PstI</option>
                                        <option value="221">PstI-HF�</option>
                                        <option value="222">PvuI-HF�</option>
                                        <option value="223">PvuII</option>
                                        <option value="224">PvuII-HF�</option>
                                        <option value="225">RsaI</option>
                                        <option value="226">RsrII</option>
                                        <option value="227">SacI-HF�</option>
                                        <option value="228">SacII</option>
                                        <option value="229">SalI</option>
                                        <option value="230">SalI-HF�</option>
                                        <option value="231">SapI</option>
                                        <option value="232">Sau3AI</option>
                                        <option value="233">Sau96I</option>
                                        <option value="234">SbfI-HF�</option>
                                        <option value="235">ScaI-HF�</option>
                                        <option value="236">ScrFI</option>
                                        <option value="237">SexAI</option>
                                        <option value="238">SfaNI</option>
                                        <option value="239">SfcI</option>
                                        <option value="240">SfiI</option>
                                        <option value="241">SfoI</option>
                                        <option value="242">SgrAI</option>
                                        <option value="243">SmaI</option>
                                        <option value="244">SmlI</option>
                                        <option value="245">SnaBI</option>
                                        <option value="246">SpeI-HF�</option>
                                        <option value="247">SphI</option>
                                        <option value="248">SphI-HF�</option>
                                        <option value="249">SrfI</option>
                                        <option value="250">SspI-HF�</option>
                                        <option value="251">StuI</option>
                                        <option value="252">StyD4I</option>
                                        <option value="253">StyI-HF�</option>
                                        <option value="254">SwaI</option>
                                        <option value="255">TaqI-v2</option>
                                        <option value="256">TfiI</option>
                                        <option value="257">TseI</option>
                                        <option value="258">Tsp45I</option>
                                        <option value="259">TspMI</option>
                                        <option value="260">TspRI</option>
                                        <option value="261">Tth111I</option>
                                        <option value="262">WarmStart� Nt.BstNBI</option>
                                        <option value="263">XbaI</option>
                                        <option value="264">XcmI</option>
                                        <option value="265">XhoI</option>
                                        <option value="266">XmaI</option>
                                        <option value="267">XmnI</option>
                                        <option value="268">ZraI</option>
                                    </select>
                                </div>
                                <div class="form-select">
                                    <label for="dire" class="required">Select sequence direction</label>
                                    <select id="dire">
                                        <option value="" disabled selected>Select sequence direction</option>
                                        <option value="F">forward</option>
                                        <option value="R">reverse</option>
                                    </select>
                                </div>
                                <div class="form-input">
                                    <label for="seq" class="required">Insert DNA Sequence</label>
                                    <input type="text" name="seq" id="seq" />
                                </div>
                            </div>
                        </div>
                        <div class="form-submit">
                            <input pys-onclick="Final" type="submit" value="submit" class="submit" id="submit" />
                        </div>
                        <div class="form-input">
                            <label for="output">OUTPUT:</label>
                            <div id="output"></div>
                        </div>
                    </form>
                    <!-- <div class="form-row">
                        <div class="form-group">
                            <div class="form-input">
                                <label for="output">OUTPUT</label>
                                <div id="output"></div>
                            </div>
                        </div>
                    </div>
                     -->

                    <py-env>
                        - pandas
                        - numpy
                    </py-env>

                    <py-script>
                        import pandas as pd
                        import numpy as np
                        from pyodide.http import open_url

                        #Read restriction enzyme data and corresponding recognition site
                        EnzymeFile = pd.read_csv(open_url('https://raw.githubusercontent.com/NanaAdwoaNewman/RecogSiteMod/main/RestrictionEnzymes.csv'))
                        res = EnzymeFile['Enzyme'].values
                        recog = EnzymeFile['Sequence'].values

                        resEnzyme = [i for i in res if i is not np.nan]
                        recogSite = [i for i in recog if i is not np.nan]

                        #AMINO ACID SEQUENCES
                        Ala = ["GCT", "GCC", "GCA", "GCG"]               #Ala
                        Arg = ["CGT", "CGC", "CGA", "CGG", "AGA", "AGG"] #Arg
                        Asn = ["AAT", "AAC"]                             #Asn
                        Asp = ["GAT", "GAC"]                             #Asp
                        Cys = ["TGT", "TGC"]                             #Cys
                        Glu = ["GAA", "GAG"]                             #Glu
                        Gln = ["CAA", "CAG"]                             #Gln
                        Gly = ["GGT", "GGC", "GGA", "GGG"]               #Gly
                        His = ["CAT", "CAC"]                             #His
                        Ile = ["ATT", "ATC", "ATA"]                      #Ile
                        Leu = ["TTA", "TTG", "CTT", "CTC", "CTA", "CTG"] #Leu
                        Lys = ["AAA","AAG"]                              #Lys
                        #Met = ["ATG"]                                    #Met
                        Phe = ["TTT", "TTC"]                             #Phe
                        Pro = ["CCT", "CCC", "CCA", "CCG"]               #Pro
                        Ser = ["TCT", "TCC", "TCA", "TCG", "AGT", "AGC"] #Ser
                        Thr = ["ACT", "ACC", "ACA", "ACG"]               #Thr
                        #Trp = ["TGG"]                                    #Trp
                        Tyr = ["TAT", "TAC"]                             #Tyr
                        Val = ["GTT", "GTC", "GTA", "GTG"]               #Val

                        #aminoAcids = np.array(['A','R','N','D','C','E','Q','G','H','I','L','K','M','F','P','S','T','W','Y','V'])
                        aminoAcidsArray = [Ala,Arg,Asn,Asp,Cys,Glu,Gln,Gly,His,Ile,Leu,Lys,Phe,Pro,Ser,Thr,Tyr,Val]

                        #BASES
                        R = ['A', 'G'] #Purine
                        Y = ['T', 'C'] #Pyrimidine
                        W = ['A', 'T']
                        S = ['G', 'C']
                        M = ['A', 'C']
                        K = ['G', 'T']
                        H = ['A', 'T', 'C']
                        B = ['G', 'C', 'T']
                        V = ['G', 'A', 'C']
                        D = ['G', 'A', 'T']
                        N = ['G', 'A', 'T', 'C']

                        bases = np.array(['R','Y','W','S','M','K','H','B','V','D'])
                        basesArray = [R, Y, W, S, M, K, H, B, V, D]


                        #Check length of DNA Sequence
                        def CheckSequenceLen(sequence):
                            lenSeq = len(sequence)
                            msg = "DNA sequence is mutated or length of DNA sequence is invalid"
                            if lenSeq%3 == 0:
                                return "done"
                            else:
                                return msg

                        #Check start and end codons 
                        def StartEndCodonCheck(sequence, direction):
                            if direction == "F":
                                startCodon = sequence[0:3].lower()
                                stopCodon = sequence[-3:].lower()
                            elif direction == "R":
                                startCodon = sequence[-3:][::-1].lower()
                                stopCodon = sequence[0:3][::-1].lower()
                                
                            if stopCodon == "tag" or stopCodon == "tga" or stopCodon == "taa":
                                if startCodon == "atg":
                                    return "1"
                                else:
                                    return "Invalid Start Codon"
                            else:
                                return "Invalid Stop Codon"
                                
                        #Checking the compatibilty of seq2modify and recogSiteCode when 'N' is present
                        #in recogSiteCode
                        def NInRecogSiteCompare(seq2modify, recogSiteCode):
                            x = 0
                            for i in recogSiteCode:
                                if seq2modify[x] == i or i == 'N' or i in bases:
                                    x = x + 1
                                else:
                                    return 0
                                    break
                            if x == len(recogSiteCode):
                                return 1

                        def reverseStr(sequence, direction):
                            if direction == 'F':
                                sequence = sequence
                            elif direction == 'R':
                                sequence = sequence[::-1]
                            return sequence
                        
                        
                        def recogSiteModification(sequence, resEnzIndex, direction):
                            sequence = reverseStr(sequence, direction)
                            resEnz = resEnzyme[resEnzIndex]
                            recogSiteCode = recogSite[resEnzIndex]
                            #print(recogSiteCode)
                            lenSequence = len(sequence)
                            lenRecogSite = len(recogSiteCode)
                            origSequence = sequence
                            Xs = []
                            #Loop through the gene sequence 
                            for x in range(lenSequence - lenRecogSite + 1):
                                recogSiteCode = recogSite[resEnzIndex]
                                seq2modify = sequence[x:x+lenRecogSite].upper()
                                for i in range(lenRecogSite):
                                    if recogSiteCode[i] == seq2modify[i]:
                                        continue
                                    else:
                                        if recogSiteCode[i] in bases:
                                            baseIndex = np.where(bases==recogSiteCode[i])[0][0]
                                            for j in basesArray[baseIndex]:
                                                if j == seq2modify[i]:
                                                    recogSiteCode = recogSiteCode[:i]+ seq2modify[i] + recogSiteCode[i+1:]
                                                    break
                                        else:
                                            break      
                                #Compare the modified recognition code with the section of the sequence         
                                if seq2modify.lower() == recogSiteCode.lower():
                                    if lenRecogSite == 3:
                                        if x%3 == 1:
                                            for c in aminoAcidsArray:
                                                    if seq2modify in c:
                                                        for d in c:
                                                            if d == seq2modify[:-1]:
                                                                continue
                                                            else:
                                                                sequence =  sequence[:x] + d + sequence[x+3:]
                                                                Xs.append(x)
                                                                break
                                                    else:
                                                        continue
                                            
                                    elif lenRecogSite == 4:
                                        #NB: * shows the start and end of the restriction enzyme recognition site
                                        #ATG|C*AG|CT*C|TAG
                                        if x%3 == 1:
                                            continue
                                        #ATG|CA*A|GCT*|TAG
                                        elif x%3 == 2:
                                            for a in aminoAcidsArray:
                                                if seq2modify[1:] in a:
                                                    for b in a:
                                                        if b == seq2modify[1:]:
                                                            continue
                                                        else:
                                                            sequence =  sequence[:x+1] + b + sequence[x+4:]
                                                            Xs.append(x+1)
                                                            break
                                                else:
                                                    continue
                                        #ATG|*AGC|T*CG|TAG
                                        else:
                                            for c in aminoAcidsArray:
                                                if seq2modify[:-1] in c:
                                                    for d in c:
                                                        if d == seq2modify[:-1]:
                                                            continue
                                                        else:
                                                            sequence =  sequence[:x] + d + sequence[x+3:]
                                                            Xs.append(x)
                                                            break
                                                else:
                                                    continue
                                    else:
                                        #ATG|C*GG|ATC*|CGA|TAG
                                        if x%3 == 1:
                                            for a in aminoAcidsArray:
                                                if seq2modify[2:5] in a:
                                                    for b in a:
                                                        if b == seq2modify[2:5]:
                                                            continue
                                                        else:
                                                            sequence =  sequence[:x+2] + b + sequence[x+5:]
                                                            Xs.append(x+2)
                                                            break
                                                else:
                                                    continue
                                        #ATG|CA*G|GAT|C*TG|TAG
                                        elif x%3 == 2:
                                            for c in aminoAcidsArray:
                                                if seq2modify[1:4] in c:
                                                    for d in c:
                                                        if d == seq2modify[1:4]:
                                                            continue
                                                        else:
                                                            sequence =  sequence[:x+1] + d + sequence[x+4:]
                                                            Xs.append(x+1)
                                                            break
                                                else:
                                                    continue
                                        #ATG|*GGA|TC*G|CGA|TAG
                                        elif x%3 == 0:
                                            for e in aminoAcidsArray:
                                                if seq2modify[:3] in e:
                                                    for f in e:
                                                        if f == seq2modify[:3]:
                                                            continue
                                                        else:
                                                            sequence =  sequence[:x] + f + sequence[x+3:]
                                                            Xs.append(x)
                                                            break
                                                else:
                                                    continue
                                #Special Case when there are Ns in the recognition site
                                elif NInRecogSiteCompare(seq2modify, recogSiteCode) == 1:
                                    subOrigSeq = sequence
                                    countOfNs = recogSiteCode.count('N')
                                    firstNPos = recogSiteCode.find('N')
                                    lastNPos = recogSiteCode.rfind('N')
                                    for i in range(lenRecogSite):
                                        if recogSiteCode[i] == seq2modify[i]:
                                            continue
                                        elif recogSiteCode[i] == 'N':
                                            recogSiteCode = recogSiteCode[:i]+ seq2modify[i] + recogSiteCode[i+1:]
                                        elif recogSiteCode[i] in bases:
                                            baseIndex = np.where(bases==recogSiteCode[i])[0][0]
                                            for j in basesArray[baseIndex]:
                                                if j == seq2modify[i]:
                                                    recogSiteCode = recogSiteCode[:i]+ seq2modify[i] + recogSiteCode[i+1:]
                                                    break
                                        else:
                                            break
                                    #print(recogSiteCode)
                                    if seq2modify == recogSiteCode:
                                        check = 0
                                        if x%3 == 0:
                                            for a in aminoAcidsArray:
                                                if seq2modify[:3] in a:
                                                    for b in a:
                                                        if b == seq2modify[:3]:
                                                            continue
                                                        else:
                                                            sequence =  sequence[:x] + b + sequence[x+3:]
                                                            check = check + 1
                                                            if firstNPos == 1:
                                                                if sequence[x] == subOrigSeq[x]:
                                                                    sequence = subOrigSeq
                                                                    check = 0
                                                                else:
                                                                    Xs.append(x)
                                                            elif firstNPos == 2:
                                                                if sequence[x:x+2] == subOrigSeq[x:x+2]:
                                                                    sequence = subOrigSeq
                                                                    check = 0
                                                                else:
                                                                    Xs.append(x)
                                                            else:
                                                                Xs.append(x)
                                                                break
                                                else:
                                                    continue
                                        elif x%3 == 1:
                                            if firstNPos > 2:
                                                for a in aminoAcidsArray:
                                                    if seq2modify[2:5] in a:
                                                        for b in a:
                                                            if b == seq2modify[2:5]:
                                                                continue
                                                            else:
                                                                sequence =  sequence[:x+2] + b + sequence[x+5:]
                                                                check = check + 1
                                                                if firstNPos == 3:
                                                                    if sequence[x+2] == subOrigSeq[x+2]:
                                                                        sequence = subOrigSeq
                                                                        check = 0
                                                                    else:
                                                                        Xs.append(x+2)
                                                                elif firstNPos == 4:
                                                                    if sequence[x+2:x+4] == subOrigSeq[x+2:x+4]:
                                                                        sequence = subOrigSeq
                                                                        check = 0
                                                                    else:
                                                                        Xs.append(x+2)
                                                                else:
                                                                    Xs.append(x+2)
                                                                    break
                                                    else:
                                                        continue
                                        elif x%3 == 2:
                                            if firstNPos >1:
                                                for a in aminoAcidsArray:
                                                    if seq2modify[1:4] in a:
                                                        for b in a:
                                                            if b == seq2modify[1:4]:
                                                                continue
                                                            else:
                                                                sequence =  sequence[:x+1] + b + sequence[x+4:]
                                                                check = check + 1
                                                                if firstNPos == 2:
                                                                    if sequence[x+1] == subOrigSeq[x+1]:
                                                                        sequence = subOrigSeq
                                                                        check = 0
                                                                    else:
                                                                        Xs.append(x+1)
                                                                elif firstNPos == 3:
                                                                    if sequence[x+1:x+3] == subOrigSeq[x+1:x+3]:
                                                                        sequence = subOrigSeq
                                                                        check = 0
                                                                    else:
                                                                        Xs.append(x+1)
                                                                else:
                                                                    Xs.append(x+1)
                                                                    break
                                                    else:
                                                        continue
                                        #looking at the end of the seq2modify in case looking at start is not fruitful
                                        if check == 0:
                                            #print(check)
                                            LastBaseIdx = x+lenRecogSite-1
                                            if (LastBaseIdx)%3 == 2:
                                                for a in aminoAcidsArray:
                                                    if seq2modify[-3:] in a:
                                                        for b in a:
                                                            if b == seq2modify[-3:]:
                                                                continue
                                                            else:
                                                                sequence =  sequence[:LastBaseIdx-2] + b + sequence[LastBaseIdx+1:]
                                                                if lenRecogSite -1 - lastNPos == 1:
                                                                    if sequence[x+lastNPos+1] == subOrigSeq[x+lastNPos+1]:
                                                                        sequence = subOrigSeq
                                                                    else:
                                                                        Xs.append(LastBaseIdx-2)
                                                                elif lenRecogSite - 1 - lastNPos == 2:
                                                                    if sequence[x+lastNPos+1:x+lastNPos+3] == subOrigSeq[x+lastNPos+1:x+lastNPos+3]:
                                                                        sequence = subOrigSeq
                                                                    else:
                                                                        Xs.append(LastBaseIdx-2)
                                                                else:
                                                                    Xs.append(LastBaseIdx-2)
                                                                    break
                                                    else:
                                                        continue
                                            elif (LastBaseIdx)%3 == 1:
                                                if lenRecogSite < 5:
                                                    continue
                                                else:
                                                    for a in aminoAcidsArray:
                                                        if seq2modify[-5:-2] in a:
                                                            for b in a:
                                                                if b == seq2modify[-5:-2]:
                                                                    continue
                                                                else:
                                                                    sequence =  sequence[:LastBaseIdx-4] + b + sequence[LastBaseIdx-1:]
                                                                    if lenRecogSite - 1 - lastNPos < 3:
                                                                        sequence = subOrigSeq
                                                                    elif lenRecogSite - 1 - lastNPos == 3:
                                                                        if sequence[x+lastNPos+1] == subOrigSeq[x+lastNPos+1]:
                                                                            sequence = subOrigSeq
                                                                        else:
                                                                            Xs.append(LastBaseIdx-4)
                                                                    elif lenRecogSite - 1 - lastNPos == 4:
                                                                        if sequence[x+lastNPos+1:x+lastNPos+3] == subOrigSeq[x+lastNPos+1:x+lastNPos+3]:
                                                                            sequence = subOrigSeq
                                                                        else:
                                                                            Xs.append(LastBaseIdx-4)
                                                                    else:
                                                                        Xs.append(LastBaseIdx-4)
                                                                        break
                                                        else:
                                                            continue
                                            elif (LastBaseIdx)%3 == 0:
                                                if lenRecogSite < 4:
                                                    continue
                                                else:
                                                    for a in aminoAcidsArray:
                                                        if seq2modify[-4:-1] in a:
                                                            for b in a:
                                                                if b == seq2modify[-4:-1]:
                                                                    continue
                                                                else:
                                                                    sequence =  sequence[:LastBaseIdx-3] + b + sequence[LastBaseIdx:]
                                                                    if lenRecogSite - 1 - lastNPos < 2:
                                                                        sequence = subOrigSeq
                                                                    elif lenRecogSite - 1 - lastNPos == 2:
                                                                        if sequence[x+lastNPos+1] == subOrigSeq[x+lastNPos+1]:
                                                                            sequence = subOrigSeq
                                                                        else:
                                                                            Xs.append(LastBaseIdx-3)
                                                                    elif lenRecogSite - 1 - lastNPos == 3:
                                                                        if sequence[x+lastNPos+1:x+lastNPos+3] == subOrigSeq[x+lastNPos+1:x+lastNPos+3]:
                                                                            sequence = subOrigSeq
                                                                        else:
                                                                            Xs.append(LastBaseIdx-3)
                                                                    else:
                                                                        Xs.append(LastBaseIdx-3)
                                                                        break
                                                        else:
                                                            continue
                            if sequence == origSequence:
                                sequence = "Could not modify sequence"
                            else:
                                for i in Xs:
                                    sequence = sequence[:i] + ' *' + sequence[i:i+3] + '* ' + sequence[i+3:]

                            sequence = reverseStr(sequence, direction)
                            #print(sequence)
                            return sequence

                        #Final
                        def Final(*args,**kwargs):
                            sequence = Element('seq').value
                            direction = Element('dire').value
                            resEnzIndex = int(Element('enzyme').value)
                            msg = Element('output')
                            #msg.write(CheckSequenceLen(sequence))

                            if CheckSequenceLen(sequence) == "done":
                                if StartEndCodonCheck(sequence, direction) == "1":
                                    msg.write(recogSiteModification(sequence, resEnzIndex, direction))
                                else:
                                    msg.write(StartEndCodonCheck(sequence, direction))
                            else:
                                msg.write(CheckSequenceLen(sequence))

                    </py-script>
                </div>
            </div>
        </div>

    </div>

    <!-- JS -->
    <!-- <script src="vendor/jquery/jquery.min.js"></script>
    <script src="vendor/nouislider/nouislider.min.js"></script>
    <script src="vendor/wnumb/wNumb.js"></script>
    <script src="vendor/jquery-validation/dist/jquery.validate.min.js"></script>
    <script src="vendor/jquery-validation/dist/additional-methods.min.js"></script>
    <script src="js/main.js"></script> -->
</body><!-- This templates was made by Colorlib (https://colorlib.com) -->
</html>